src/ts/app.tsx

```diff
@@ -0,0 +1,187 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import { List, Map, is } from 'immutable';
+
+interface Config {
+  origin: string;
+  apiRoot: string;
+  token: string;
+  status: TokenStatus;
+}
+
+enum TokenStatus {
+  unchecked, verified, failed
+}
+
+interface OptionProps {
+  saveConfigs: (configs: List<Config>) => void;
+  configs: List<Config>;
+  github: boolean;
+}
+
+interface OptionState {
+  configs: List<Config>;
+  newGithub: string;
+}
+
+const TokenStatusButton: React.StatelessComponent<{status: TokenStatus} & React.HTMLProps<HTMLButtonElement>> = (props) => {
+  const {status, ...otherProps} = props;
+  const text = ((status) => {
+    switch (status) {
+      case TokenStatus.verified:  return 'verified';
+      case TokenStatus.unchecked: return 'connection test';
+      case TokenStatus.failed:    return 'failed';
+    }
+  })(status);
+  return <button {...otherProps}>{text}</button>;
+}
+
+class App extends React.PureComponent<OptionProps, OptionState> {
+  constructor(props: OptionProps) {
+    super(props);
+    this.state = {
+      configs: List(props.configs),
+      newGithub: ''
+    };
+
+    this.addItem = this.addItem.bind(this);
+    this.removeItem = this.removeItem.bind(this);
+    this.updateStatus = this.updateStatus.bind(this);
+    this.saveConfigs = this.saveConfigs.bind(this);
+    this.onConfigChanged = this.onConfigChanged.bind(this);
+  }
+
+  addItem() {
+    const newGithub = this.state.newGithub;
+    const apiRoot = (newGithub === 'https://github.com') ? 'https://api.github.com' : (newGithub + '/api/v3')
+    if (newGithub !== '' && this.state.configs.filter(({origin}) => (origin === newGithub)).isEmpty) {
+      let config: Config = { origin: newGithub, apiRoot: apiRoot, token: '', status: TokenStatus.unchecked }
+      this.setState(({configs}) => ({
+        configs: configs.push(config),
+        newGithub: '',
+      }));
+    }
+  }
+
+  removeItem(i: number) {
+    this.setState(({configs}) => ({
+      configs: configs.delete(i)
+    }));
+  }
+
+  updateStatus(i: number) {
```

Comment:
I think `updateTokenStatus` is better, because `status` and `this.state` is confusingly similar.